File: medium/2075-decode-the-slanted-ciphertext/solution.py
class Solution:
    def decodeCiphertext(self, encodedText: str, rows: int) -> str:
        """
        2075. Decode the Slanted Ciphertext
        Time: O(n)
        Space: O(n)
        """
        if not encodedText:
            return ""

        cols = len(encodedText) // rows
        result = []

        for start_col in range(cols):
            row, col = 0, start_col
            while row < rows and col < cols:
                result.append(encodedText[row * cols + col])
                row += 1
                col += 1

        return ''.join(result).rstrip()

File: medium/2075-decode-the-slanted-ciphertext/solution.java
/**
 * 2075. Decode the Slanted Ciphertext
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public String decodeCiphertext(String encodedText, int rows) {
        if (encodedText.isEmpty()) {
            return "";
        }

        int cols = encodedText.length() / rows;
        StringBuilder result = new StringBuilder();

        for (int startCol = 0; startCol < cols; startCol++) {
            int row = 0, col = startCol;
            while (row < rows && col < cols) {
                result.append(encodedText.charAt(row * cols + col));
                row++;
                col++;
            }
        }

        int end = result.length();
        while (end > 0 && result.charAt(end - 1) == ' ') {
            end--;
        }

        return result.substring(0, end);
    }
}
